manageiq-operator/pkg/helpers/miq-components/httpd.go

```diff
@@ -278,197 +303,195 @@ func initializeHttpdContainer(spec *miqv1alpha1.ManageIQSpec, privileged bool, c
 			Name:      "httpd",
 			Namespace: cr.ObjectMeta.Namespace,
-			Labels:    deploymentLabels,
 		},
 		Spec: appsv1.DeploymentSpec{
-			Replicas: &repNum,
 			Selector: &metav1.LabelSelector{
-				MatchLabels: podLabels,
+				MatchLabels: deploymentLabels,
 			},
 			Template: corev1.PodTemplateSpec{
 				ObjectMeta: metav1.ObjectMeta{
+					Labels: deploymentLabels,
 					Name:   "httpd",
-					Labels: podLabels,
-				},
-				Spec: corev1.PodSpec{
-					Containers: []corev1.Container{container},
-					Volumes: []corev1.Volume{
-						corev1.Volume{
-							Name: "httpd-config",
-							VolumeSource: corev1.VolumeSource{
-								ConfigMap: &corev1.ConfigMapVolumeSource{
-									LocalObjectReference: corev1.LocalObjectReference{Name: "httpd-configs"},
-								},
-							},
-						},
-					},
 				},
+				Spec: corev1.PodSpec{},
 			},
 		},
 	}
 
-	configureHttpdAuth(&cr.Spec, &deployment.Spec.Template.Spec)
-
 	// Only assign the service account if we need additional privileges
 	if privileged {
 		deployment.Spec.Template.Spec.ServiceAccountName = cr.Spec.AppName + "-httpd"
 	}
 
-	return deployment, nil
+	f := func() error {
+		if err := controllerutil.SetControllerReference(cr, deployment, scheme); err != nil {
+			return err
+		}
+		addAppLabel(cr.Spec.AppName, &deployment.ObjectMeta)
+		var repNum int32 = 1
+		deployment.Spec.Replicas = &repNum
+		deployment.Spec.Template.Spec.Containers = []corev1.Container{container}
+		deployment.Spec.Template.Spec.Volumes = []corev1.Volume{
+			corev1.Volume{
+				Name: "httpd-config",
+				VolumeSource: corev1.VolumeSource{
+					ConfigMap: &corev1.ConfigMapVolumeSource{
+						LocalObjectReference: corev1.LocalObjectReference{Name: "httpd-configs"},
+					},
+				},
+			},
+		}
+		configureHttpdAuth(&cr.Spec, &deployment.Spec.Template.Spec)
+		return nil
+	}
+
+	return deployment, f, nil
 }
 
-func NewUIService(cr *miqv1alpha1.ManageIQ) *corev1.Service {
-	labels := map[string]string{
-		"app": cr.Spec.AppName,
-	}
-	selector := map[string]string{
-		"service": "ui",
-	}
-	return &corev1.Service{
+func UIService(cr *miqv1alpha1.ManageIQ, scheme *runtime.Scheme) (*corev1.Service, controllerutil.MutateFn) {
+	service := &corev1.Service{
 		ObjectMeta: metav1.ObjectMeta{
 			Name:      "ui",
 			Namespace: cr.ObjectMeta.Namespace,
-			Labels:    labels,
 		},
 		Spec: corev1.ServiceSpec{
-			Ports: []corev1.ServicePort{
-				corev1.ServicePort{
-					Name: "web-service-3000",
-					Port: 3000,
-				},
-			},
-			Selector: selector,
+			Ports: []corev1.ServicePort{},
 		},
 	}
-}
 
-func NewWebService(cr *miqv1alpha1.ManageIQ) *corev1.Service {
-	labels := map[string]string{
-		"app": cr.Spec.AppName,
-	}
-	selector := map[string]string{
-		"service": "web-service",
+	f := func() error {
+		if err := controllerutil.SetControllerReference(cr, service, scheme); err != nil {
+			return err
+		}
+		addAppLabel(cr.Spec.AppName, &service.ObjectMeta)
+		service.Spec.Ports[0].Name = "ui-service-3000"
```

Comment:
Same as https://github.com/ManageIQ/manageiq-pods/pull/577/files#r450876638, I think we need to check for nil as that's the array zero value.

I'm not sure if the `Get` in `CreateOrUpdate` will overwrite the Ports array, but it's probably safer to remove it from the spec above and `make` a new one here if it's nil

Reply:
This really applies to all of the other services as well. I'll only write it here though.

Reply:
Done